dashboard/modules/job/tests/test_job_manager.py

```diff
@@ -459,5 +455,107 @@ def test_stop_job_subprocess_cleanup_upon_stop(self, job_manager):
             wait_for_condition(check_subprocess_cleaned, pid=pid)
 
 
+class TestTailLogs:
+    @pytest.mark.asyncio
+    async def test_unknown_job(self, job_manager):
+        with pytest.raises(
+                RuntimeError, match="Job 'unknown' does not exist."):
+            async for _ in job_manager.tail_job_logs("unknown"):
+                pass
+
+    @pytest.mark.asyncio
+    async def test_successful_job(self, job_manager):
+        """Test tailing logs for a PENDING -> RUNNING -> SUCCESSFUL job."""
+        start_signal_actor = SignalActor.remote()
+
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            _, tmp_file, job_id = _run_hanging_command(
+                job_manager, tmp_dir, start_signal_actor=start_signal_actor)
+
+            # TODO(edoakes): check we get no logs before actor starts (not sure
+            # how to timeout the iterator call).
+            assert job_manager.get_job_status(
+                job_id).status == JobStatus.PENDING
+
+            # Signal job to start.
+            ray.get(start_signal_actor.send.remote())
+
+            i = 0
+            async for lines in job_manager.tail_job_logs(job_id):
+                print(lines, end="")
+                assert all(
+                    s == "Waiting..." for s in lines.strip().split("\n"))
+                if i == 5:
+                    break
+                i += 1
+
+            # Signal the job to exit by writing to the file.
+            with open(tmp_file, "w") as f:
+                print("hello", file=f)
+
+            async for lines in job_manager.tail_job_logs(job_id):
+                assert all(
+                    s == "Waiting..." for s in lines.strip().split("\n"))
+                print(lines, end="")
```

Comment:
Shouldn't we have more checks here for succeeded / stopped / failed cases for things like # of "Waiting ..." returned and final log message near the end to differentiate them ? In current tests i don't think the subsequent polling gives us any additional signal since 1) we always only prints "Waiting ... " in logs 2) we don't check how many lines of "Waiting ..." we got after making a job state transition call, it could be many or none 3) we don't print or check anything at the end of the script to ensure we indeed get the last line as expected.

Reply:
How about we have a script that should print like 

```
Waiting ...
Waiting ...
Waiting ...
Waiting ...
Waiting ... (We called stop)
Stop received, waiting ...
Stop received, waiting ...
Stopped !
```

So it covers PENDING, RUNNING logs, ability to make job changes while polling logs, subsequent polling does not produce more / fewer lines as expected and we ensured we got the very last line of log in all cases ?